rust/src/array_data.rs

```diff
@@ -0,0 +1,270 @@
+    length: i64,
+
+    /// The number of null elements in this array data
+    null_count: i64,
+
+    /// The offset into this array data.
+    offset: i64,
+
+    /// The buffers for this array data. Note that depending on the array types, this
+    /// could hold different kinds of buffers (e.g., value buffer, value offset buffer)
+    /// at different positions.
+    buffers: Vec<Buffer>,
+
+    /// The child(ren) of this array. Only non-empty for nested types, currently
+    /// `ListArray` and `StructArray`.
+    child_data: Vec<ArrayDataRef>,
+
+    /// The null bitmap. A `None` value for this indicates all values are non-null in
+    /// this array.
+    null_bitmap: Option<Bitmap>,
+}
+
+pub type ArrayDataRef = Arc<ArrayData>;
+pub const UNKNOWN_NULL_COUNT: i64 = -1;
+
+impl ArrayData {
+    pub fn new(
+        data_type: DataType,
+        length: i64,
+        mut null_count: i64,
+        null_bit_buffer: Option<Buffer>,
+        offset: i64,
+        buffers: Vec<Buffer>,
+        child_data: Vec<ArrayDataRef>,
+    ) -> Self {
+        if null_count < 0 {
+            null_count = if let Some(ref buf) = null_bit_buffer {
+                count_set_bits(buf.data())
+            } else {
+                0
+            }
+        }
+        let null_bitmap = null_bit_buffer.map(Bitmap::from);
+        Self {
+            data_type,
+            length,
+            null_count,
+            offset,
+            buffers,
+            child_data,
+            null_bitmap,
+        }
+    }
+
+    /// Returns a builder to construct a `ArrayData` instance.
+    pub fn builder(data_type: DataType) -> ArrayDataBuilder {
+        ArrayDataBuilder::new(data_type)
+    }
+
+    /// Returns a copy of the data type for this array data.
+    pub fn data_type(&self) -> DataType {
+        self.data_type.clone()
+    }
+
+    /// Returns a reference to the data type of this array data.
+    pub fn data_type_ref(&self) -> &DataType {
+        &self.data_type
+    }
+
+    /// Returns a slice of buffers for this array data.
+    pub fn buffers(&self) -> &[Buffer] {
+        &self.buffers[..]
+    }
+
+    /// Returns a slice of children data arrays.
+    pub fn child_data(&self) -> &[ArrayDataRef] {
+        &self.child_data[..]
+    }
+
+    /// Returns whether the element at index `i` is null.
+    pub fn is_null(&self, i: usize) -> bool {
+        if let Some(ref b) = self.null_bitmap {
+            return !b.is_set(i);
+        }
+        false
+    }
+
+    /// Returns a reference to the null bitmap of this array data.
+    pub fn null_bitmap(&self) -> &Option<Bitmap> {
+        &self.null_bitmap
+    }
+
+    /// Returns whether the element at index `i` is not null.
+    pub fn is_valid(&self, i: usize) -> bool {
+        if let Some(ref b) = self.null_bitmap {
+            return b.is_set(i);
+        }
+        true
+    }
+
+    /// Returns the length (i.e., number of elements) of this array.
+    pub fn length(&self) -> usize {
```

Comment:
a) Rust uses `len` for container sizes, so I think that's what we should call it.
b) why is the internal datatype `i64` and you cast to `usize` here? Let's say we don't support 16bit systems and length hopefully does not contain negative values (which you don't check in `ArrayDataBuilder::len`, on both 32 and 64 bit this conversion can never fail). So why not storing `usize` in the first place?

Reply:
Good point. I think we should use `len` and don't return `usize` here. 

I followed the C++ `ArrayData` structure and used `i64` here, but it seems Arrow format specify that Array length should be stored as a 32-bit signed integer. @wesm : could you clarify why `int64_t` is used for C++ `ArrayData` length - seems `int32_t` is enough?

Because of the format specification, we can't use `usize` to store the length. Also, to my understanding Array is a logical concept and could contain one or more physical Buffers, therefore the length could go beyond the limit of `usize`(e.g., you can have a 32-bit long Array on a 16-bit platform).